libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+    _polyfence_numpoints = 0;
+
+    // for each point on polygon fence
+    // Note: boundary is "unclosed" meaning the last point is *not* the same as the first
+    for (uint8_t i=0; i<num_points; i++) {
+
+        // find points before and after current point (relative to current point)
+        const uint8_t before_idx = (i == 0) ? num_points-1 : i-1;
+        Vector2f before_pt = boundary[before_idx] - boundary[i];
+        Vector2f after_pt = boundary[i+1] - boundary[i];
+
+        // if points are overlapping fail
+        if (before_pt.is_zero() || after_pt.is_zero() || (before_pt == after_pt)) {
+            return false;
+        }
+
+        // scale points to be unit vectors
+        before_pt.normalize();
+        after_pt.normalize();
+
+        // calculate intermediate point and scale to margin
+        Vector2f intermediate_pt = (after_pt + before_pt) * 0.5f;
+        float intermediate_len = intermediate_pt.length();
+        intermediate_pt *= (margin_cm / intermediate_len);
+
+        // find final point which is inside the original polygon
+        _polyfence_pts[i] = boundary[i] + intermediate_pt;
+        if (Polygon_outside(_polyfence_pts[i], boundary, num_points)) {
+            _polyfence_pts[i] = boundary[i] - intermediate_pt;
+            if (Polygon_outside(_polyfence_pts[i], boundary, num_points)) {
+                // could not find a point on either side that was within the fence so fail
+                // this can happen if fence lines are closer than margin_m
+                return false;
+            }
+        }
+
+        _polyfence_numpoints++;
+    }
+
+    // last point is same as first point
+    _polyfence_pts[_polyfence_numpoints] = _polyfence_pts[0];
+    _polyfence_numpoints++;
+
+    // record fence update time so we don't process this exact fence again
+    _polyfence_update_ms = fence_ptr->get_boundary_update_ms();
+
+    return true;
+}
+
+// create a visibility graph of the polygon fence
+// returns true on success
+// requires create_polygon_fence_with_margin to have been run
+bool AP_OADijkstra::create_polygon_fence_visgraph()
+{
+    // exit immediately if no polygon fence (with margin)
+    if (_polyfence_numpoints == 0) {
+        return false;
+    }
+
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
+        return false;
+    }
+
+    // get polygon boundary
+    uint16_t num_points;
+    const Vector2f* boundary = fence_ptr->get_boundary_points(num_points);
+    if ((boundary == nullptr) || (num_points < 3)) {
+        return false;
+    }
+
+    // clear polygon visibility graph
+    _polyfence_visgraph.clear();
+
+    // calculate distance from each polygon fence point to all other points
+    for (uint8_t i=0; i<_polyfence_numpoints-1; i++) {
+        for (uint8_t j=i+1; j<_polyfence_numpoints-1; j++) {
+            const Vector2f &start1 = _polyfence_pts[i];
```

Comment:
Take this reference outside the inner loop so you can reuse it, without having to recalculate it all the time.

Reply:
nice, thanks.